lib/graphql/dataloader/background_loader.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+begin
+  require "concurrent"
+rescue LoadError
+  raise LoadError, "GraphQL::Dataloader::BackgroundLoader requires concurrent-ruby, add `gem \"concurrent-ruby\"` to your gemfile, `bundle install`, then try again!"
+end
+
+module GraphQL
+  class Dataloader
+    # A loader whose `#perform` will run on a background thread.
+    #
+    # It uses `Concurrent::Future` which uses a global thread pool.
+    class BackgroundLoader < GraphQL::Dataloader::Loader
+      def sync
+        Concurrent::Future.execute do
+          setup_thread
+          super
+          teardown_thread
+        end
+      end
+
+      # Implement this method to prepare any thread state that your
+      # application depends on, for example, `Thread.current[:user] = @context[:current_user]`
```

Comment:
I do not follow how this is helpful. If this method is running in a separate thread anyways, why would the thread not access `@context[:current_user]` directly?

Reply:
I've worked on applications which depend on context like 

```ruby
class User 
  def self.current 
    Thread.current[:user]
  end 

  def self.current=(u)
    Thread.current[:user] = u 
  end 
end 
```

And I wanted to support a case like that. (Just for legacy code purposes.) But in fact, it should probably go in `#perform` anyways, so I've removed those methods!